compat: Add atexit()

ThePiGrepper opened the pull request:
- static implementation of atexit() with ATEXIT_MAX=32 registered
  functions.

I wrote this yesterday so I thought it could be useful after seeing that an issue was opened for it. It's a simple static solution. Expanding it to one that allocates memory would be relatively simple, although I don't know if it would be actually needed. It might solve #1816 .

anton-bondarev commented:
@ThePiGrepper This issue has been already assigned with @sk2sat. Can you work together? It's rather difficult task.

anton-bondarev commented:
@ThePiGrepper Thanks
You move to the right way. But this solution is not good, because maybe there are a lot of application in Embox. And it's necessary to separate different tasks (processes).
I think You should look at task resources and how implemented pthread_cancel && pthread_key and so on

ThePiGrepper commented:
I see. I'll look into that. Also, regarding having a solution with memory allocation or having a static minimal solution. which one would you prefer? maybe both as different modules and have the internals written into atexit_api(I believe that's the name of the empty module already there)

alexkalmuk commented:
> I see. I'll look into that. Also, regarding having a solution with memory allocation or having a static minimal solution. which one would you prefer? maybe both as different modules and have the internals written into atexit_api(I believe that's the name of the empty module already there)

I agree with @anton-bondarev that each task should have its own atexit table, which is most likely pre-allocated when a task is created. But in case if you have one single statically preallocated atexit table you claim that there is at most one task in the system, but even in that case, this table can be allocated for this task, not globally.

ThePiGrepper commented:
I just updated the PR using the new task resource. @anton-bondarev , please review it when you have time.

anton-bondarev commented on src/compat/posix/proc/atexit/atexit.c:
```diff
@@ -0,0 +1,38 @@
+
+void __atexit_funcs(void)
+{
+	void (*atexit_func)(void);
```
You can use 
list = task_self_resource_atexit();

anton-bondarev commented on src/compat/posix/proc/atexit/atexit.c:
```diff
@@ -0,0 +1,38 @@
+
+struct task_atexit *list;
+
+static int slot;
```
You have not to use global variables.
Update 'struct task_atexit' adding int slot, and use if for all date

anton-bondarev commented on src/compat/posix/proc/atexit/atexit.c:
```diff
@@ -0,0 +1,38 @@
+}
+
+int atexit(void (*func)(void)) {
+	if (list == NULL) {
```
don't use global variables just
list = task_self_resource_atexit();

anton-bondarev commented:
@ThePiGrepper Thanks, but you don't use global variables, use only task variables
Task resources allocate memory on the top of the task stack, and after that, it can be used for the local variables in the task

ThePiGrepper commented:
I understand. yeah, I was familiarizing myself with the task resources code. the init handle gets called always before going to the process' main then, right? if that's the case(which I was having doubts about), indeed, I could just do  list = task_self_resource_atexit(); and must add all the other global related variables(ie:slot) as part of the task_atexit struct.
thanks for the review.
updated patch in a bit.

ThePiGrepper commented:
> @ThePiGrepper Thanks, but you don't use global variables, use only task variables
> Task resources allocate memory on the top of the task stack, and after that, it can be used for the local variables in the task

done.

anton-bondarev commented on src/kernel/task/resource/atexit/atexit.h:
```diff
@@ -14,6 +14,7 @@
 
 struct task_atexit {
 	void (*atexit_func[ATEXIT_MAX])(void);
+	size_t slot;
```
i think that it's better to use simple types if there is not necessary in the opposite, i think it's enough to use 'int' there

ThePiGrepper replied:
done

anton-bondarev commented on src/compat/posix/proc/exit/exit.c:
```diff
@@ -33,7 +33,14 @@ void _exit(int status) {
 }
 
+/* from atexit */
+void __attribute__((weak)) __atexit_funcs(void)
```
why do you want to use the weak function there?

ThePiGrepper replied:
indeed, good idea. changed

anton-bondarev commented:
@ThePiGrepper Thanks, but please don't rebase during code view, just commit ahead

How are you testing that atexit() work correctly?

ThePiGrepper commented:
> @ThePiGrepper Thanks, but please don't rebase during code view, just commit ahead
> 
> How are you testing that atexit() work correctly?

first, I run I wrote a single test to see if it worked at all, however that one was missing a way to check the task context. now, I added a second command which fork/execs several times the first one, each time with different number of atexit() being called, including 0, and then check if each time the number of atexit being called is the correct one.

ThePiGrepper commented:
btw, on a semi-related issue, how are you dealing with static storage duration variables (either global or static local variables)? are you zeroing them before each call?

anton-bondarev commented:
> first, I wrote a single test to see if it worked at all, however that one was missing a way to check the task context. now, I added a second command which fork/execs several times the first one, each time with different number of atexit() being called, including 0, and then checks if each time the number of atexit being called is the correct one.

Great! Could you please create a unit test with this functionalities?
I want to reproduce one before merging!

anton-bondarev commented on src/compat/posix/proc/exit/exit.c:
```diff
@@ -33,7 +33,11 @@ void _exit(int status) {
+
 /* stdlib */
 void exit(int status) {
+	__atexit_funcs();
```
You should also add to Mybuild depends from atexit_api where __atexit_funcs is implemented

ThePiGrepper replied:
done.

ThePiGrepper commented:
> > first, I wrote a single test to see if it worked at all, however that one was missing a way to check the task context. now, I added a second command which fork/execs several times the first one, each time with different number of atexit() being called, including 0, and then checks if each time the number of atexit being called is the correct one.
> 
> Great! Could you please create a unit test with this functionalities?
> I want to reproduce one before merging!



> > first, I wrote a single test to see if it worked at all, however that one was missing a way to check the task context. now, I added a second command which fork/execs several times the first one, each time with different number of atexit() being called, including 0, and then checks if each time the number of atexit being called is the correct one.
> 
> Great! Could you please create a unit test with this functionalities?
> I want to reproduce one before merging!

ok, will do.

anton-bondarev commented:
@ThePiGrepper Great. It works for me!
But some comment for the test:

- use test_emit() and then test_assert_emitted() for checking sequence instead of  printf()
- make static bye() routine. It's used only local.

P.S. Please don't rebase during a code review, just commit ahead your branch

ThePiGrepper commented:
> @ThePiGrepper Great. It works for me!
> But some comment for the test:
> 
>     * use test_emit() and then test_assert_emitted() for checking sequence instead of  printf()
> 
>     * make static bye() routine. It's used only local.
> 
> 
> P.S. Please don't rebase during a code review, just commit ahead your branch

done.
if everything's ok, I'd like to rebase and squash some commit before you merge. thx.

anton-bondarev commented:
@ThePiGrepper I've approved PR. Now please rebase under embox/master and squash fixups

ThePiGrepper commented:
> @ThePiGrepper I've approved PR. Now please rebase under embox/master and squash fixups

done. thx
